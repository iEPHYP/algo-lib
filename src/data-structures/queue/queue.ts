import { EventEmitter } from 'events';

import { LinkedList } from '../linked-list';

export class Queue<TValue = number> {
  public linkedList = new LinkedList<TValue>();

  public toList() {
    return this.linkedList.toList();
  }

  public enqueue(value: TValue) {
    this.linkedList.push(value);
    this.eventEmitter.emit('enqueue', value);
  }

  public dequeue() {
    const dequeuedItem = this.linkedList.shift();
    this.eventEmitter.emit('dequeue', dequeuedItem);

    return dequeuedItem;
  }

  public empty() {
    return this.linkedList.empty();
  }

  public size() {
    return this.linkedList.size();
  }

  public clear() {
    this.linkedList.clear();
  }

  private eventEmitter = new EventEmitter();

  public subscribeToEnqueue(callback: (value: TValue) => void) {
    this.eventEmitter.on('enqueue', callback);
  }

  public subscribeToDequeue(callback: (value: ReturnType<typeof this.dequeue>) => void) {
    this.eventEmitter.on('enqueue', callback);
  }
}
