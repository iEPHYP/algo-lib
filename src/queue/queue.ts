import { LinkedList } from '../linked-list';

export class Queue<TValue = number> {
  public linkedList = new LinkedList<TValue>();

  public toList() {
    return this.linkedList.toList();
  }

  public enqueue(value: TValue) {
    this.linkedList.push(value);
  }

  public dequeue() {
    return this.linkedList.shift();
  }

  public empty() {
    return this.linkedList.empty();
  }

  public size() {
    return this.linkedList.size();
  }

  public clear() {
    this.linkedList.clear();
  }
}
