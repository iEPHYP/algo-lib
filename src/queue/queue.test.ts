import { Queue } from './queue';

describe('Queue', () => {
  let queue: Queue;

  beforeEach(() => {
    queue = new Queue();
  });

  it('initializes correctly', () => {
    expect(queue.toList()).toMatchSnapshot('initialized Queue in List');
  });

  it('enqueues elements', () => {
    queue.enqueue(1);
    expect(queue.toList()).toMatchSnapshot('on empty Queue');
    queue.enqueue(2);
    expect(queue.toList()).toMatchSnapshot('on Queue with 1 item');
    queue.enqueue(3);
    expect(queue.toList()).toMatchSnapshot('on Queue with 2 items');
  });

  it('dequeues elements', () => {
    expect(queue.dequeue()).toBe(null);
    expect(queue.toList()).toMatchSnapshot('on empty Queue');
    queue.enqueue(1);
    expect(queue.dequeue()).toBe(1);
    expect(queue.toList()).toMatchSnapshot('on Queue with 1 item');
    queue.clear();
    queue.enqueue(1);
    queue.enqueue(2);
    expect(queue.dequeue()).toBe(1);
    expect(queue.toList()).toMatchSnapshot('on Queue with 2 items');
    queue.clear();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    expect(queue.dequeue()).toBe(1);
    expect(queue.toList()).toMatchSnapshot('on Queue with 3 items');
  });

  it('checks if queue is empty', () => {
    expect(queue.empty()).toBe(true);
    queue.enqueue(1);
    expect(queue.empty()).toBe(false);
    queue.dequeue();
    expect(queue.empty()).toBe(true);
  });

  it('checks the size of the queue', () => {
    expect(queue.size()).toBe(0);
    queue.enqueue(1);
    expect(queue.size()).toBe(1);
    queue.enqueue(2);
    expect(queue.size()).toBe(2);
    queue.dequeue();
    expect(queue.size()).toBe(1);
  });

  it('performs combined operations', () => {
    queue.enqueue(1);
    queue.enqueue(2);
    expect(queue.dequeue()).toBe(1);
    queue.enqueue(3);
    expect(queue.dequeue()).toBe(2);
    expect(queue.toList()).toMatchSnapshot();
  });
});
