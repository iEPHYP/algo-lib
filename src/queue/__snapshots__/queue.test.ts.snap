// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue dequeues elements: on Queue List with 1 item 1`] = `[]`;

exports[`Queue dequeues elements: on Queue List with 2 items 1`] = `
[
  2,
]
`;

exports[`Queue dequeues elements: on Queue List with 3 items 1`] = `
[
  2,
  3,
]
`;

exports[`Queue dequeues elements: on Queue with 1 item 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 0,
    "head": null,
    "tail": null,
  },
}
`;

exports[`Queue dequeues elements: on Queue with 2 items 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 1,
    "head": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": null,
        "value": 1,
      },
      "value": 2,
    },
    "tail": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": null,
        "value": 1,
      },
      "value": 2,
    },
  },
}
`;

exports[`Queue dequeues elements: on Queue with 3 items 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 2,
    "head": {
      "nextItem": {
        "nextItem": null,
        "previousItem": [Circular],
        "value": 3,
      },
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": null,
        "value": 1,
      },
      "value": 2,
    },
    "tail": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": {
          "nextItem": [Circular],
          "previousItem": null,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
  },
}
`;

exports[`Queue dequeues elements: on empty Queue 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 0,
    "head": null,
    "tail": null,
  },
}
`;

exports[`Queue dequeues elements: on empty Queue List 1`] = `[]`;

exports[`Queue enqueues elements: on Queue List with 1 item 1`] = `
[
  1,
  2,
]
`;

exports[`Queue enqueues elements: on Queue List with 2 items 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`Queue enqueues elements: on Queue with 1 item 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 2,
    "head": {
      "nextItem": {
        "nextItem": null,
        "previousItem": [Circular],
        "value": 2,
      },
      "previousItem": null,
      "value": 1,
    },
    "tail": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": null,
        "value": 1,
      },
      "value": 2,
    },
  },
}
`;

exports[`Queue enqueues elements: on Queue with 2 items 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 3,
    "head": {
      "nextItem": {
        "nextItem": {
          "nextItem": null,
          "previousItem": [Circular],
          "value": 3,
        },
        "previousItem": [Circular],
        "value": 2,
      },
      "previousItem": null,
      "value": 1,
    },
    "tail": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": {
          "nextItem": [Circular],
          "previousItem": null,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
  },
}
`;

exports[`Queue enqueues elements: on empty Queue 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 1,
    "head": {
      "nextItem": null,
      "previousItem": null,
      "value": 1,
    },
    "tail": {
      "nextItem": null,
      "previousItem": null,
      "value": 1,
    },
  },
}
`;

exports[`Queue enqueues elements: on empty Queue List 1`] = `
[
  1,
]
`;

exports[`Queue initializes correctly: initialized Queue 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 0,
    "head": null,
    "tail": null,
  },
}
`;

exports[`Queue initializes correctly: initialized Queue in List 1`] = `[]`;

exports[`Queue performs combined operations 1`] = `
Queue {
  "linkedList": LinkedList {
    "_size": 1,
    "head": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": {
          "nextItem": [Circular],
          "previousItem": null,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
    "tail": {
      "nextItem": null,
      "previousItem": {
        "nextItem": [Circular],
        "previousItem": {
          "nextItem": [Circular],
          "previousItem": null,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
  },
}
`;

exports[`Queue performs combined operations 2`] = `
[
  3,
]
`;
